lib/vsdata.cpp

```diff
@@ -47,6 +47,104 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int dim = mesh->Dimension();
+   int ref = 1;
+   if (dim == 3)
+   {
+      for (int i = 0; i < mesh->GetNBE(); i++)
+      {
+         const int f = mesh->GetBdrElementFaceIndex(i);
+         const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+         const FiniteElement *fe = NULL;
+         if (fec && fec->GetContType() != FiniteElementCollection::DISCONTINUOUS)
+         {
+            fe = gf.FESpace()->GetFaceElement(f);//might abort instead of returning NULL!
+         }
+         int order;
+         if (fe)
+         {
+            order = fe->GetOrder();
+            if (fe->GetMapType() == FiniteElement::MapType::INTEGRAL)
+            {
+               order += mesh->GetBdrElementTransformation(i)->OrderW();
+            }
+         }
+         else
+         {
+            int ivol, info;
+            mesh->GetBdrElementAdjacentElement(i, ivol, info);
+            const FiniteElement *fevol = gf.FESpace()->GetFE(ivol);
+            order = fevol->GetOrder() - 1;
+            if (fevol->GetMapType() == FiniteElement::MapType::INTEGRAL)
+            {
+               order += mesh->GetElementTransformation(ivol)->OrderW();
+            }
+         }
+         ref = std::max(ref, 2 * order);
+      }
+   }
+   else
+   {
+      for (int i = 0; i < mesh->GetNE(); i++)
+      {
+         const FiniteElement &fe = *gf.FESpace()->GetFE(i);
+         int order = fe.GetOrder();
+         if (fe.GetMapType() == FiniteElement::MapType::INTEGRAL)
+         {
+            order += mesh->GetElementTransformation(i)->OrderW();
+         }
+         ref = std::max(ref, 2 * order);
+      }
+   }
+
+   return ref;
+}
+
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor()
+{
+   if (!sol) { return 1; }
+
+   return (sol->Size() == mesh->GetNV())?(2):(1);
```

Comment:
Why return 2 if `sol` is just vertex data? In what case is `sol->Size()` not the same as `mesh->GetNV()`?

Reply:
Vertex data = linear function = refinement 2 * 1 = 2. The check is maybe not necessary, I just saw something like that somewhere else in the code.

Reply:
Is not it an option to load cell-centered data? Or maybe was that an intention which was then dropped 🤔

Reply:
Hmm, setting order_ref for linear elements to be 2 ... it is consistent with the multiplication by 2 in the other function, however, I'm not convinced that's what we want. I'd prefer to use `order_ref=order`, not `order_ref=2*order`.

I don't think we have an option to load piecewise constant data (cell-centered) that is not a grid-function. Unless someone added that and I did not see it. Where did you see a comparison like `sol->Size() == mesh->GetNV()`?

Reply:
Ok, I changed the formula to `2*order-1`, which should be sufficient and allows to simplify the logic here, so vertex data will not be refined 😉 .

Reply:
I think it is more than sufficient to use simply `order_ref=order` -- this will indicate that there is curvature without refining too much; any additional refinements, if desired, will be up to the user.